File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.py
import heapq

def get_time(time):
    return int(time[:2]) * 60 + int(time[3:])

def solution(book_time):
    book_time.sort();
    
    rooms = [get_time(book_time[0][1])]
    for i in range(1, len(book_time)):
        if rooms[0] + 10 <= get_time(book_time[i][0]):
            heapq.heappop(rooms)
        heapq.heappush(rooms, get_time(book_time[i][1]))
    
    return len(rooms)

File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <queue>

using namespace std;

int get_time(string time){
    int hour = stoi(time.substr(0, 2));
    int minute = stoi(time.substr(3));
    
    return hour * 60 + minute;
}

int solution(vector<vector<string>> book_time) {
    // 입실 시간을 기준으로 오름차순 정렬
    sort(book_time.begin(), book_time.end(), [](vector<string> a, vector<string> b){ return a[0] < b[0]; });
    
    priority_queue<int, vector<int>, greater<int>> rooms; // min_heap
    rooms.push(get_time(book_time[0][1]));
    for(int i = 1; i < book_time.size(); i++){
        int start_time = get_time(book_time[i][0]);
        int end_time = get_time(book_time[i][1]);
        
        if(rooms.top() + 10 <= start_time) rooms.pop();
        
        rooms.push(end_time);
    }
    
    return rooms.size();
}